cpp/src/arrow/flight/types_async.h

```diff
@@ -0,0 +1,116 @@
+
+/// \brief Non-templated state for an async RPC.
+class AsyncListenerBase {
+ public:
+  AsyncListenerBase();
+  virtual ~AsyncListenerBase();
+
+  /// \brief Request cancellation of the RPC.
+  ///
+  /// The RPC is not cancelled until AsyncListener::OnFinish is called.
+  void TryCancel();
+
+ private:
+  friend class arrow::flight::internal::ClientTransport;
+  friend class arrow::flight::IpcPutter;
+
+  /// Transport-specific state for this RPC.  Transport
+  /// implementations may store and retrieve state here via
+  /// ClientTransport::SetAsyncRpc and ClientTransport::GetAsyncRpc.
+  std::unique_ptr<internal::AsyncRpc> rpc_state_;
+};
+
+/// \brief Callbacks for results from async RPCs.
+///
+/// A single listener may not be used for multiple concurrent RPC
+/// calls.  The application MUST hold the listener alive until
+/// OnFinish() is called and has finished.
+template <typename T>
+class ARROW_FLIGHT_EXPORT AsyncListener : public AsyncListenerBase {
+ public:
+  /// \brief Get the next server result.
+  /// This will never be called concurrently with itself or OnFinish.
+  virtual void OnNext(T message) = 0;
+  /// \brief Get the final status.
+  /// This will never be called concurrently with itself or OnNext.
+  virtual void OnFinish(TransportStatus status) = 0;
+};
+
+/// \brief Callbacks for results from async RPCs that read Arrow data.
+class ARROW_FLIGHT_EXPORT IpcListener : public AsyncListener<FlightStreamChunk> {
+ public:
+  /// \brief Get the IPC schema.
+  /// This will never be called concurrently with itself, OnNext, or OnFinish.
+  virtual void OnSchema(std::shared_ptr<Schema> schema) = 0;
+};
+
+/// \brief Callbacks for DoPut.
+class ARROW_FLIGHT_EXPORT IpcPutter : public AsyncListener<std::unique_ptr<Buffer>> {
+ public:
+  // TODO: which of these can be made const T&?
+
+  /// \brief Begin writing an IPC stream.  May only be called once.
+  ///   Must be called before writing any record batches.
+  void Begin(std::shared_ptr<Schema> schema,
+             ipc::IpcWriteOptions options = ipc::IpcWriteOptions::Defaults());
+  /// \brief Write a record batch.
+  void Write(std::shared_ptr<RecordBatch> batch) {
+    Write({std::move(batch), NULLPTR}, false);
+  }
+  /// \brief Write application metadata.  May be called before Begin.
+  void Write(std::shared_ptr<Buffer> app_metadata) {
+    Write({NULLPTR, std::move(app_metadata)}, false);
+  }
+  /// \brief Write a record batch with application metadata.
+  ///
+  /// \param[in] last If true, this is the last write on this stream.
+  ///   This may let the underlying transport optimize the write.
+  ///   Semantically equivalent to calling Write then DoneWriting.
+  void Write(FlightStreamChunk chunk, bool last = false);
+  /// \brief Indicate that the client is done writing.  Must be
+  ///   called, or OnFinish will never be called.
+  void DoneWriting();
+
+  /// \brief Begin or Write finished.  The application may Write() again.
+  virtual void OnWritten() = 0;
```

Comment:
So the application cannot issue several `Write` calls if the corresponding `OnWritten` notifications haven't been received yet?

Reply:
Right, only one Write can be 'active' at any time.

Reply:
But why not let the caller buffer writes?

Reply:
We could do that (the underlying API does not, but that's OK), but then we probably need to consider backpressure.

Reply:
Indeed... but backpressure is precisely the kind of concern that the transport is better equipped to deal with, no?

Reply:
The `PauseProducing`/`ResumeProducing` mechanism might be an inspiration.

Reply:
Not here I think (it's not possible to implement with gRPC, and is that really a proven mechanism? Did Acero ever actually implement it or was it only a theoretical API)

Reply:
That API is modeled after [Twisted's](https://docs.twisted.org/en/stable/core/howto/producers.html#push-producers) flow-control mechanism.

Another possibility is to expose a [asyncio-like](https://docs.python.org/3.13/library/asyncio-stream.html#asyncio.StreamWriter.drain) API, for example a `Future<> Drain()` method on the transport.

Reply:
I am going to implement queueing, at least.

I think Start/StopProducing could be implemented on top of just knowing the queue depth.

gRPC already has backpressure, anyways: it's "one write" at a time but the write is considered complete once it enters an internal buffer, so effectively gRPC is giving you backpressure by delaying the OnWrite callback.

Reply:
But Flight-level queueing is also useful to smooth over the fact that one "write" may actually be multiple gRPC-level writes

Reply:
If all transports give you the queue depth, then why not?
But the open question is how the application would know what to do with the queue depth. It's not easy to interpret.

Reply:
Well, but only having Start/StopProducing means baking in the logic into a slower-moving library instead.  

For queue depth: on each OnWritten callback, the application would check the queue depth and enqueue messages until there's enough work in the buffer. If it wants to be more sophisticated, it can track (or we could expose) an estimate of the bytes enqueued instead. If you really want to be sophisticated, you could estimate bandwidth by also tracking the time between OnWritten callbacks.

For both gRPC and UCX, we'd have to bake some variant of this logic into the library.

Reply:
> For queue depth: on each OnWritten callback, the application would check the queue depth and enqueue messages until there's enough work in the buffer.

1) doesn't the transport have better knowledge to decide whether there is "enough work in the buffer"?

2) if the application decides to postpone the next write, to which deadline does it postpone it? does it fire a timer to check the queue depth, say, 1ms later?

The point of having something like `Drain` or `ResumeProducing` is that the _transport_ informs the application that there is now enough room in the buffer.

Reply:
1: well, yes, that's why gRPC doesn't call the OnWritten callback until it's ready for more work!

2: the application would just do the check the next time OnWritten is called (it's called for *each* message sent). So you can think of OnWritten as ResumeProducing, if you want.

Reply:
> the application would just do the check the next time OnWritten is called (it's called for _each_ message sent). So you can think of OnWritten as ResumeProducing, if you want.

Ah! Well, fair enough, then.

Reply:
I'll try to clarify that (clearly this is confusing/not obvious)